### ISSUE TITLE
JWKS verification failing

### ISSUE THREAD

**Issue by JamieCressey** [micronaut-projects/micronaut-security#50]
I have two apps running locally, an auth server;

```
micronaut:
  application:
    name: auth
  security:
    enabled: true
    endpoints:
      login:
        enabled: true
        path: auth/login
      oauth:
        enabled: true
        path: auth/oauth
      keys:
        enabled: true
        path: auth/keys
    token:
      jwt:
        enabled: true
```

Which implements a custom `JwkProvider` and `RSASignatureGeneratorConfiguration`;

```
@Singleton
public class RotatingJwksProvider implements JwkProvider {

    @Inject
    KeyPairRepository keyPairRepository;


    @Override
    public List<JWK> retrieveJsonWebKeys() {
        List<JWK> retval = new ArrayList<>();
        List<KeyPair> keyPairs = keyPairRepository.getAll();
        for (KeyPair keyPair : keyPairs) {
            RSAPublicKey rsaPublicKey = keyPair.getRSAPublicKey();

            RSAKey jwk = new RSAKey.Builder(rsaPublicKey)
                    .keyID(keyPair.getId())
                    .keyUse(KeyUse.SIGNATURE)
                    .algorithm(JWSAlgorithm.RS256)
                    .build();
            retval.add(jwk.toPublicJWK());

        }
        return retval;
    }

}
```

```
@Singleton
@Named("generator")
public class RotatingRSASignatureConfiguration implements RSASignatureGeneratorConfiguration {

    private final KeyPairRepository keyPairRepository;
    private static KeyPair keyPair = null;

    @Inject
    public RotatingRSASignatureConfiguration(KeyPairRepository keyPairRepository) {
        this.keyPairRepository = keyPairRepository;
    }

    private synchronized void ensureKeyPair() {
        if (keyPair == null) {
            keyPair = keyPairRepository.getLatest();
        }
        if (keyPair == null) {
            keyPair = keyPairRepository.save(KeyPair.factory());
        }
    }

    @Override
    public RSAPublicKey getPublicKey() {
        ensureKeyPair();
        return keyPair.getRSAPublicKey();
    }

    @Override
    public RSAPrivateKey getPrivateKey() {
        ensureKeyPair();
        return keyPair.getRSAPrivateKey();
    }

    @Override
    public JWSAlgorithm getJwsAlgorithm() {
        return JWSAlgorithm.RS256;
    }
}
```

And a client server;

```
micronaut:
  server:
    port: 8081
  security:
    enabled: true
    endpoints:
      login:
        enabled: false
      oauth:
        enabled: false
    token:
      jwt:
        enabled: true
        signatures:
          jwks:
            auth:
              url: "http://localhost:8080/auth/keys"

```

I can login to the auth server and receive a JWT token;

```
curl -X POST \
  http://localhost:8080/auth/login \
  -H 'Content-Type: application/json' \
  -d '{
    "username": "jamie",
    "password": "xxx"
}'

{"username":"jamie","roles":["admintestrole"],"access_token":"xxx","refresh_token":"xxx","token_type":"Bearer","expires_in":3600}
```

And I have successfully verified the generated JWT (and public key) using jwt.io;

```
Header:

{
  "alg": "RS256"
}

Payload:

{
  "sub": "jamie",
  "nbf": 1559404743,
  "roles": [
    "admintestrole"
  ],
  "iss": "auth",
  "exp": 1559408343,
  "iat": 1559404743
}
```
However when I attempt to use that access token against the client service, it fails;

```
curl -v http://localhost:8081/jctest   -H 'Authorization: Bearer xxx'   -H 'Content-Type: application/json'
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8081 (#0)
> GET /jctest HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.54.0
> Accept: */*
> Authorization: Bearer xxx
> Content-Type: application/json
> 
< HTTP/1.1 401 Unauthorized
< Date: Sat, 1 Jun 2019 17:02:25 GMT
< transfer-encoding: chunked
< connection: close
< 
* Closing connection 0

```

And the following is displayed in the logs;

```
2019-06-01 17:52:05  DEBUG JwtTokenValidator:130 - JWT is signed
2019-06-01 17:52:05  DEBUG JwtTokenValidator:154 - No algorithms are supported
2019-06-01 17:52:05  DEBUG JwtTokenValidator:159 - No signature algorithm found for JWT: xxx
```

I have also confirmed the client service is accessing the keys endpoint of the auth service;

```
2019-06-01 17:14:48  DEBUG SecurityFilter:157 - Authorized request GET /auth/keys. The rule provider io.micronaut.security.rules.SecuredAnnotationRule authorized the request.
2019-06-01 17:14:48  DEBUG RoutingInBoundHandler:1307 - Encoding emitted response object [{"keys":[{"kty":"RSA","e":"AQAB","use":"sig","kid":"b4224a75-a902-46ea-a4fc-23d5b14153e2","alg":"RS256","n":"1uZ7e3YPcbHFDwMSlm8CN6WH4hpA6VxIoNdfiyPL49Vww8DrInzpBDFnGlQ9kwg-5Dzif_fGXF3BI4Uxzs-lyB3GLNmNPx-MDfZL63d7PE-Ydcj4Y6xEy6siH2IyXK3WFY-s_fX8ewT1Gam0rvvcFkvuOA0voCOrnX92-LvcdGDtJbsdFskI60JLfRNiVab21T8BsrY4xw8qZeZwDgUBQyVEz38wLOi2aXy9wTI7jIDhBweEOcl6Jz2nFymEModRbdLc5gBrQ6E6KPLfiR0KaXjWuAUjhh06iN9_ureKloFQiYp3CKUcDnEhnuIr53cz42SId06cPTEvZPcplo7BIQ"}]}] using codec: io.micronaut.runtime.http.codec.TextPlainCodec@1e30a519
```

Is there something I have missed?

Thanks!

**Comment by gausnes** [comment]
I ran into a similar issue, but it was the reserve scenario. The algorithm header wasn't being emitted in the `jwks` from our auth server, which blew up.

I was able to ask our security team to emit the `alg` header with the respective keys to fix this issue. I feel like the default behavior maybe should changed for this case as the `alg` header is optional.

Not sure about the `JWT` spec, but it feels like micronaut is taking an overly strict interpretation of the spec.
